lib/grpc/server/interceptors/cors.ex

```diff
@@ -0,0 +1,98 @@
+defmodule GRPC.Server.Interceptors.CORS do
+  @moduledoc """
+  Sends CORS headers when the client is calling the rpc via web transcoding or grpcweb.
+
+  ## Options
+
+    * `allow_origin` - A string containing the allowed origin(s), or a remote function (e.g. `&MyApp.MyModule.function/2)`) which takes a `req` and a `stream` and returns a string. Defaults to `"*"`, which will allow all origins to access this endpoint.
+    * `allow_headers` - A string containing the allowed headers, or a remote function (e.g. `&MyApp.MyModule.function/2)`) which takes a `req` and a `stream` and returns a string. Defaults to the value of the `"access-control-request-headers"` request header from the client.
+
+  ## Usage
+
+      defmodule Your.Endpoint do
+        use GRPC.Endpoint
+
+        intercept GRPC.Server.Interceptors.CORS
+      end
+
+      defmodule Your.Endpoint do
+        use GRPC.Endpoint
+
+        intercept GRPC.Server.Interceptors.CORS, allow_origin: "some.origin"
+      end
+
+
+      defmodule Your.Endpoint do
+        use GRPC.Endpoint
+
+        def allow_origin(req, stream), do: "calculated.origin"
+        intercept GRPC.Server.Interceptors.CORS, allow: &Your.Endpoint.allow_origin/2
+      end
+  """
+
+  @behaviour GRPC.Server.Interceptor
+  @impl true
+  def init(opts \\ []) do
+    # the funky first clause matches a 2-arity remote is_function
+    # note that this function is run in the context of a macro, which brings some limitations with it
+    allow_origin =
+      case Keyword.get(opts, :allow_origin) do
+        {:&, [], [{:/, [], [_signature, 2]}]} = fun -> fun
+        static when is_binary(static) -> static
+        _ -> "*"
+      end
```

Comment:
As this is running in a macro context (I can't recall why, but let's roll with this), the current code does not allow for a variable to be referenced here, I think. It will end up defaulting to the catch-all silently and this is a security risk.

Perhaps this validation should be done at run time?

Reply:
Pass an atom/integer/tuple/list/variable referencing a binary instead of a capture or literal binary and this will fall through to "*".

Reply:
We should raise with ArgumentError

Reply:
> Pass an atom/integer/tuple/list/variable referencing a binary instead of a capture or literal binary and this will fall through to "*".

This is actually covered in the documentation: a string or a remote function, otherwise is defaults to '*'. As was the default values used.

It's a moot point now, in any case, as I've removed the default for `allow_origin` altogether in  18ad11f69d38a4b75e312ba617b1d59e93ba5216 which should address this and another of your concerns.